lib/puppet/defaults.rb

```diff
@@ -1354,17 +1354,21 @@ def self.initialize_default_settings!(settings)
       executable.",
     },
     :serverport => {
-      :default    => "8140",
+      :default    => 8140,
       :desc       => "The default port puppet subcommands use to communicate
       with Puppet Server. (eg `puppet facts upload`, `puppet agent`). May be
       overridden by more specific settings (see `ca_port`, `report_port`).",
     },
     :masterport => {
       :default    => "$serverport",
+      :type       => :port,
       :desc       => "The default port puppet subcommands use to communicate
       with Puppet Server. (eg `puppet facts upload`, `puppet agent`). May be
       overridden by more specific settings (see `ca_port`, `report_port`).",
-      :hook => proc { |value| Puppet[:serverport] = value }
+      :hook => proc do |value|
```

Comment:
This probably needs to be `call_hook => :on_initialize_and_write` in case `masterport/serverport` is defined in the `[agent]` section, see https://github.com/puppetlabs/puppet/pull/8395

Reply:
I think `:hook` not being properly evaluated (regardless of call_hook) when a setting is in a section other than `[main]` or `[user]` is a bug. Left a more detailed comment here: https://github.com/puppetlabs/puppet/pull/8411#discussion_r509964361